### PR TITLE
Refactoring keg_relocate to use ELFShim#interpreter 

### PR DESCRIPTION
**Pull request by rmNULL**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

Replaces all uses  of `ELFShim#with_interpreter?` and removes it.
Introduce  `ELFShim#interpreter` to return the interpreter in path 
if present otherwise return nil.

@sjackman @woodruffw  @MikeMcQuaid 
there are style issues and concerns parsing interpreter from the output of shell command,
these are discussed in review comments.

### CODE REVIEW

File: Library/Homebrew/extend/os/linux/keg_relocate.rb [hunk 1]
```diff
@@ -41,15 +41,13 @@ def change_rpath(file, old_prefix, new_prefix)
     new_rpath = rpath.join(":")
     cmd = [patchelf, "--force-rpath", "--set-rpath", new_rpath]
 
-    if file.with_interpreter?
-      old_interpreter = Utils.safe_popen_read(patchelf, "--print-interpreter", file).strip
-      new_interpreter = if File.readable? "#{new_prefix}/lib/ld.so"
+    old_interpreter = file.interpreter
+    new_interpreter = old_interpreter && if File.readable? "#{new_prefix}/lib/ld.so"
```
Comment (rmNULL) [inline Library/Homebrew/extend/os/linux/keg_relocate.rb:45]: rubocop doesn't accept this, how do i clean this, be verbose?

File: Library/Homebrew/os/linux/elf.rb [hunk 2]
```diff
@@ -68,28 +68,30 @@ def binary_executable?
...
+        nil
       end
+    elsif (patchelf = DevelopmentTools.locate "patchelf")
+      # the check is probably redundant, it is already done by #keg_relocate
+      # adding it as a safety precaution. And maybe for something else to use
+      # in future.
+      interp = Utils.popen_read(patchelf, "--print-interpreter", to_s, err: out).strip
+      interp.start_with?("patchelf:") ? nil : interp
+    elsif (file = DevelopmentTools.locate("file"))
+      interp = Utils.popen_read(file, "-L", "-b", to_s, err: out).strip
```
Comment (rmNULL) [inline Library/Homebrew/os/linux/elf.rb:88]: parsing the output of `file` is troublesome.
Initially i thought of extracting text after "interpreterter " until we come across ".so,"
the problem with this approach is we can't guarantee the interpreter they picked won't have these characters in middle( e.g: `/lib/h.so,ull.so`) it may sound strange yet its a legal name to have.

File: Library/Homebrew/extend/os/linux/keg_relocate.rb (hunk 1 above)
Comment (sjackman) [reply Library/Homebrew/extend/os/linux/keg_relocate.rb:45]:
```suggestion
    old_interpreter = file.interpreter
    if old_interpreter.present?
       new_interpreter = if File.readable? "#{new_prefix}/lib/ld.so"
```

File: Library/Homebrew/extend/os/linux/keg_relocate.rb [hunk 3]
```diff
@@ -41,15 +41,13 @@ def change_rpath(file, old_prefix, new_prefix)
...
         "#{new_prefix}/lib/ld.so"
       else
         old_interpreter.sub old_prefix, new_prefix
       end
-      cmd << "--set-interpreter" << new_interpreter if old_interpreter != new_interpreter
-    end
+    cmd << "--set-interpreter" << new_interpreter if old_interpreter != new_interpreter
```
Comment (sjackman) [inline Library/Homebrew/extend/os/linux/keg_relocate.rb:50]: Please revert this change.

File: Library/Homebrew/os/linux/elf.rb [hunk 4]
```diff
@@ -68,28 +68,30 @@ def binary_executable?
...
+      begin
+        patchelf_patcher.interpreter
+      rescue PatchELF::PatchError => e
+        opoo e unless e.to_s.start_with? "No interpreter found"
+        nil
       end
+    elsif (patchelf = DevelopmentTools.locate "patchelf")
+      # the check is probably redundant, it is already done by #keg_relocate
+      # adding it as a safety precaution. And maybe for something else to use
+      # in future.
```
Comment (sjackman) [inline Library/Homebrew/os/linux/elf.rb:84]: You can remove this comment.

File: Library/Homebrew/os/linux/elf.rb (hunk 2 above)
Comment (sjackman) [reply Library/Homebrew/os/linux/elf.rb:88]:
```
$ file -b /bin/ls
ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 2.6.32, BuildID[sha1]=d0bc0fb9b3f60f72bbad3c5a1d24c9e2a1fde775, stripped
```

It's a valid concern that the interpreter may contain a comma, and thank you for raising it, but it's also pretty unlikely. The text after `, interpreter ` and before the next comma is the interpreter.

You could split the string on `, ` (comma space) and extract the element that begins with `interpreter `.

File: Library/Homebrew/extend/os/linux/keg_relocate.rb (hunk 1 above)
Comment (rmNULL) [reply Library/Homebrew/extend/os/linux/keg_relocate.rb:45]: i sense that `new_interpreter` won't defined every time. it may cause problem in below line where we check `old_interpreter != new_interpreter`

File: Library/Homebrew/os/linux/elf.rb (hunk 2 above)
Comment (rmNULL) [reply Library/Homebrew/os/linux/elf.rb:88]: alright, i'll leave a comment for future. so we are aware of it.

File: Library/Homebrew/os/linux/elf.rb (hunk 4 above)
Comment (rmNULL) [reply Library/Homebrew/os/linux/elf.rb:84]: oops, that was meant to be discussed here.  sure i'll remove the comment, but should we do that check ?

File: Library/Homebrew/os/linux/elf.rb (hunk 2 above)
Comment (sjackman) [reply Library/Homebrew/os/linux/elf.rb:88]: If you split on `, ` (comma space), I think the chances of the interpreter filename containing the string `, ` (comma space) is close to nil.

File: Library/Homebrew/os/linux/elf.rb (hunk 4 above)
Comment (sjackman) [reply Library/Homebrew/os/linux/elf.rb:84]: Yeah, this `interpreter` could potentially be called from a different location other than `keg_relocate` that doesn't guarantee that `patchelf` is installed.

File: Library/Homebrew/os/linux/elf.rb (hunk 2 above)
Comment (sjackman) [reply Library/Homebrew/os/linux/elf.rb:88]: So in that case, it's probably not even worth mentioning it in a comment.

Comment (woodruffw) [reply Library/Homebrew/os/linux/elf.rb:88]: Agreed with @sjackman here.

File: Library/Homebrew/extend/os/linux/keg_relocate.rb [hunk 5]
```diff
@@ -41,8 +41,9 @@ def change_rpath(file, old_prefix, new_prefix)
     new_rpath = rpath.join(":")
     cmd = [patchelf, "--force-rpath", "--set-rpath", new_rpath]
 
-    if file.with_interpreter?
-      old_interpreter = Utils.safe_popen_read(patchelf, "--print-interpreter", file).strip
+    old_interpreter = file.interpreter
+    new_interpreter = old_interpreter
```
Comment (sjackman) [inline Library/Homebrew/extend/os/linux/keg_relocate.rb:45]:
```suggestion
```
Please remove this line.

File: Library/Homebrew/os/linux/elf.rb [hunk 6]
```diff
@@ -68,28 +68,26 @@ def binary_executable?
...
-          opoo e unless e.to_s.start_with? "No interpreter found"
-          false
-        end
-      elsif which "readelf"
-        Utils.popen_read("readelf", "-l", to_path).include?(" INTERP ")
-      elsif which "file"
-        Utils.popen_read("file", "-L", "-b", to_path).include?(" interpreter ")
-      else
-        raise "Please install either readelf (from binutils) or file."
+    if HOMEBREW_PATCHELF_RB
```
Comment (sjackman) [inline Library/Homebrew/os/linux/elf.rb:74]:
```suggestion
    @interpreter ||= if HOMEBREW_PATCHELF_RB
```

File: Library/Homebrew/os/linux/elf.rb [hunk 7]
```diff
@@ -68,28 +68,26 @@ def binary_executable?
     elf_type == :executable
   end
 
-  def with_interpreter?
-    return @with_interpreter if defined? @with_interpreter
+  def interpreter
+    return @interpreter if defined? @interpreter
```
Comment (sjackman) [inline Library/Homebrew/os/linux/elf.rb:72]:
```suggestion
```
Please remove this line.

File: Library/Homebrew/os/linux/elf.rb [hunk 8]
```diff
@@ -68,28 +68,26 @@ def binary_executable?
...
+        nil
       end
+    elsif (patchelf = DevelopmentTools.locate "patchelf")
+      interp = Utils.popen_read(patchelf, "--print-interpreter", to_s, err: :out).strip
+      interp.start_with?("patchelf:") ? nil : interp
+    elsif (file = DevelopmentTools.locate("file"))
+      output = Utils.popen_read(file, "-L", "-b", to_s, err: :out).strip
+      return unless output.start_with?("ELF")
+
+      output[/, interpreter (.+?), /, 1]
```
Comment (sjackman) [inline Library/Homebrew/os/linux/elf.rb:88]:
```console
$ brew ruby -e 'p "ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 2.6.32, BuildID[sha1]=d0bc0fb9b3f60f72bbad3c5a1d24c9e2a1fde775, stripped"[/, interpreter (.+?), /, 1]'
"/lib64/ld-linux-x86-64.so.2"
```
Nice! I didn't know about `.+?`. That's super useful!

Comment (sjackman) [reply Library/Homebrew/os/linux/elf.rb:88]:
```suggestion
      output[/^ELF.*, interpreter (.+?), /, 1]
```

File: Library/Homebrew/os/linux/elf.rb [hunk 9]
```diff
@@ -68,28 +68,26 @@ def binary_executable?
...
+      rescue PatchELF::PatchError => e
+        opoo e unless e.to_s.start_with? "No interpreter found"
+        nil
       end
+    elsif (patchelf = DevelopmentTools.locate "patchelf")
+      interp = Utils.popen_read(patchelf, "--print-interpreter", to_s, err: :out).strip
+      interp.start_with?("patchelf:") ? nil : interp
+    elsif (file = DevelopmentTools.locate("file"))
+      output = Utils.popen_read(file, "-L", "-b", to_s, err: :out).strip
+      return unless output.start_with?("ELF")
```
Comment (sjackman) [inline Library/Homebrew/os/linux/elf.rb:86]:
```suggestion
```
Please delete this line.

File: Library/Homebrew/os/linux/elf.rb [hunk 10]
```diff
@@ -68,28 +68,26 @@ def binary_executable?
...
+
+      output[/, interpreter (.+?), /, 1]
     else
-      false
+      raise "Please either install patchelf or file"
```
Comment (sjackman) [inline Library/Homebrew/os/linux/elf.rb:90]:
```suggestion
      raise "Please either install either patchelf or file."
```

**Comment by sjackman** [comment]
This PR drops support for getting the interpreter using `readelf`. I'm okay with that though.
```
$ readelf -l /bin/ls
…
  INTERP         0x0000000000000238 0x0000000000400238 0x0000000000400238
                 0x000000000000001c 0x000000000000001c  R      1
      [Requesting program interpreter: /lib64/ld-linux-x86-64.so.2]
```

File: Library/Homebrew/extend/os/linux/keg_relocate.rb (hunk 5 above)
Comment (rmNULL) [reply Library/Homebrew/extend/os/linux/keg_relocate.rb:45]: it is for safety purpose, 

consider the case when  `old_intepreter.present? ` is `false`
as `new_interpreter` is conditionally defined the line #55 will fail with undefined error

Comment (sjackman) [reply Library/Homebrew/extend/os/linux/keg_relocate.rb:45]: Hmm. Good catch. In that case, I suggest…
```suggestion
    new_interpreter = nil
```

File: Library/Homebrew/os/linux/elf.rb (hunk 6 above)
Comment (sjackman) [reply Library/Homebrew/os/linux/elf.rb:74]: My previous `||=` suggestion was wrong because `nil` is falsey.
```suggestion
    @interpreter = if HOMEBREW_PATCHELF_RB
```

File: Library/Homebrew/extend/os/linux/keg_relocate.rb (hunk 5 above)
Comment (rmNULL) [reply Library/Homebrew/extend/os/linux/keg_relocate.rb]: i went with explicit check in the conditional below,
that explains the intention better than holding proxy values.

**Comment by rmNULL** [comment]
should we be adding tests for this ?

**Comment by rmNULL** [comment]
would we need tests for this method ?

File: Library/Homebrew/extend/os/linux/keg_relocate.rb [hunk 11]
```diff
@@ -41,8 +41,8 @@ def change_rpath(file, old_prefix, new_prefix)
     new_rpath = rpath.join(":")
     cmd = [patchelf, "--force-rpath", "--set-rpath", new_rpath]
 
-    if file.with_interpreter?
-      old_interpreter = Utils.safe_popen_read(patchelf, "--print-interpreter", file).strip
+    old_interpreter = file.interpreter
+    if old_interpreter.present?
```
Comment (sjackman) [inline Library/Homebrew/extend/os/linux/keg_relocate.rb:45]:
```suggestion
    old_interpreter = file.interpreter
    new_interpreter = if old_interpreter.nil?
      nil
    elsif File.readable? "#{new_prefix}/lib/ld.so"
      "#{new_prefix}/lib/ld.so"
    else
      old_interpreter.sub old_prefix, new_prefix
    end
    cmd << "--set-interpreter" << new_interpreter if old_interpreter != new_interpreter
```
⚠️ This code is untested.

File: Library/Homebrew/extend/os/linux/keg_relocate.rb [hunk 12]
```diff
@@ -51,7 +51,7 @@ def change_rpath(file, old_prefix, new_prefix)
       cmd << "--set-interpreter" << new_interpreter if old_interpreter != new_interpreter
     end
 
-    return if old_rpath == new_rpath && old_interpreter == new_interpreter
+    return if old_rpath == new_rpath && old_interpreter.present? && old_interpreter == new_interpreter
```
Comment (sjackman) [inline Library/Homebrew/extend/os/linux/keg_relocate.rb:54]: If `old_intepreter` is `nil`, the `patchelf` command will always run, even when `RPATH` are identical, which is not the desired behaviour.

File: Library/Homebrew/os/linux/elf.rb [hunk 13]
```diff
@@ -68,28 +68,24 @@ def binary_executable?
...
+    @interpreter = if HOMEBREW_PATCHELF_RB
+      begin
+        patchelf_patcher.interpreter
+      rescue PatchELF::PatchError => e
+        opoo e unless e.to_s.start_with? "No interpreter found"
+        nil
       end
+    elsif (patchelf = DevelopmentTools.locate "patchelf")
+      interp = Utils.popen_read(patchelf, "--print-interpreter", to_s, err: :out).strip
+      interp.start_with?("patchelf:") ? nil : interp
```
Comment (sjackman) [inline Library/Homebrew/os/linux/elf.rb:83]:
```suggestion
      $CHILD_STATUS.success? ? interp : nil
```

**Comment by sjackman** [comment]
>> This PR drops support for getting the interpreter using readelf. I'm okay with that though.

> How often is this being used by users? Can we figure this out and how disruptive it'll be before removing the code? There's not much code there and it seems weird to drop it unless it's unused/causing problems.

I believe this code is never currently used, because this `interpreter` method is called only by `keg_relocate`, which guarantees that `patchelf` is installed, and so `patchelf` will be used instead of `readelf`. That said, it's easy enough to implement `readelf`, so may as well.

@rmNULL Parse this line using a regex.
```
$ readelf -l /bin/ls
 [Requesting program interpreter: /lib64/ld-linux-x86-64.so.2]
```

**Comment by rmNULL** [comment]
> > This PR drops support for getting the interpreter using `readelf`. I'm okay with that though.
> 
> How often is this being used by users? Can we figure this out and how disruptive it'll be before removing the code? There's not much code there and it seems weird to drop it unless it's unused/causing problems.

@MikeMcQuaid 
there was no much code because it was just checking for presence of interpreter section.
however here, we have to parse that text from the output, which is a little arguable more code.


Another case, (as of now) this method is called only once(throughout the whole source), 
and patchelf is guaranteed to exist when that call is made.
patchelf was a requirement before the change too,  in that way we are not dropping any support.

**Comment by rmNULL** [comment]
> > This PR drops support for getting the interpreter using `readelf`. I'm okay with that though.
> 
> How often is this being used by users? Can we figure this out and how disruptive it'll be before removing the code? There's not much code there and it seems weird to drop it unless it's unused/causing problems.

@MikeMcQuaid 
there was no much code because it was just checking for presence of interpreter section.
however here, we have to parse that text from the output, which is a little arguable more code.


Another case, (as of now) this method is called only once(throughout the whole source), 
and patchelf is guaranteed to exist when that call is made.
patchelf was a requirement before the change too,  in that way we are not dropping any support.

File: Library/Homebrew/extend/os/linux/keg_relocate.rb (hunk 12 above)
Comment (rmNULL) [reply Library/Homebrew/extend/os/linux/keg_relocate.rb]: i've modified it to 
` return if old_rpath == new_rpath && (old_interpreter.blank? || old_interpreter == new_interpreter)`


but somewhere inside i feel, 
`return if old_interpreter.blank?  ||  (old_rpath == new_rpath && old_interpreter == new_interpreter)`
reads better.

File: Library/Homebrew/extend/os/linux/keg_relocate.rb [hunk 14]
```diff
@@ -51,7 +52,7 @@ def change_rpath(file, old_prefix, new_prefix)
       cmd << "--set-interpreter" << new_interpreter if old_interpreter != new_interpreter
     end
 
-    return if old_rpath == new_rpath && old_interpreter == new_interpreter
+    return if old_rpath == new_rpath && (old_interpreter.blank? || old_interpreter == new_interpreter)
```
Comment (sjackman) [inline Library/Homebrew/extend/os/linux/keg_relocate.rb:55]: Please revert this line. The above suggestion handles this case.

File: Library/Homebrew/extend/os/linux/keg_relocate.rb (hunk 12 above)
Comment (sjackman) [reply Library/Homebrew/extend/os/linux/keg_relocate.rb]: Please revert this line. The above suggestion handles this case.

**Comment by woodruffw** [comment]
Thanks, @rmNULL 🎉
